src/EditorFeatures/Core/IDecompiledSourceService.cs

```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis.Editor
+{
+    internal interface IDecompiledSourceService : ILanguageService
+    {
+        /// <summary>
+        /// Generates formatted source code containing general information about the symbol's
+        /// containing assembly and the decompiled source code which the given ISymbol is or is a part of
+        /// into the given document
+        /// </summary>
+        /// <param name="document">The document to generate source into</param>
+        /// <param name="symbol">The symbol to generate source for</param>
+        /// <param name="cancellationToken">To cancel document operations</param>
+        /// <returns>The updated document</returns>
+        Task<Document> AddSourceToAsync(Document document, ISymbol symbol, CancellationToken cancellationToken = default);
```

Comment:
see previous comment about cancellationTokens not being optiona.

Reply:
Ok.  So i had to read this signature a few times, and look at the impl.  It is def a bit wonky.  What we should be doing *instead* is the following:

```c#
Task<Document> DecompileAsync(Project project, ISymbol symbol, CancellationToken cancellationToken);
```

Importantly, there is no reason for you to take in a Document.  You are really only doing that so you can get the containing project to get things like the Compilation and other Language services.